### ISSUE TITLE
Simple LDAP Proxy to re-map memberOf as an attribute of inetOrgPerson

### ISSUE THREAD

**Issue by drew504** [TremoloSecurity/MyVirtualDirectory#35]
Hi Marc, All,

Thanks for the project, I think it is going to be a life saver for me. That said, I'm having difficulty deploying what I think is a very simple MyVirtualDirectory configuration. Hopefully you can show me the error of my ways.

I have an application that will only read attributes from the inetOrgPerson objectclass. I would like to leverage the memberOf attribute within my application, but in my directory (389/FreeIPA) memberOf lives in the inetUser objectclass. As a result, memberOf is invisible to my application.

My thinking is that I could use My VirtualDirectory as an LDAP proxy, using the VirtualMemberOf insert to create the memberOf attribute within inetOrgPerson.

In practice, I'm having two problems:
1) "Missing schema location in RootDSE, using default schema." - First thing I did was configure a simple proxy following the directions. It works, but when I connect with Apache Directory Studio, I get the above error. I'm not sure if this is benign or indicative of a larger issue. - I tried different configurations with the RootDSE and RootObject inserts. Neither seemed to have any impact.

2) When I include the VirtualMemberOf insert in my namespace chain, the application will not start and prints the following in the logs: 
"[2017-10-09 16:25:37,656][main] ERROR Server - Error starting server : java.lang.IllegalArgumentException: Attribute value cannot be null
java.lang.IllegalArgumentException: Attribute value cannot be null
        at com.novell.ldap.LDAPAttribute.addValue(LDAPAttribute.java:227) ~[jldap-1.0.0.jar:?]
        at net.sourceforge.myvd.inserts.mapping.VirtualMemberOf.configure(VirtualMemberOf.java:90) ~[myvd-server-1.0.1.jar:?]
        at net.sourceforge.myvd.core.InsertChain.configureChain(InsertChain.java:192) ~[myvd-server-1.0.1.jar:?]
        at net.sourceforge.myvd.server.ServerCore.configureChain(ServerCore.java:80) ~[myvd-server-1.0.1.jar:?]
        at net.sourceforge.myvd.server.ServerCore.buildNamespaces(ServerCore.java:150) ~[myvd-server-1.0.1.jar:?]
        at net.sourceforge.myvd.server.ServerCore.startService(ServerCore.java:169) ~[myvd-server-1.0.1.jar:?]
        at net.sourceforge.myvd.server.Server.startServer(Server.java:259) ~[myvd-server-1.0.1.jar:?]
        at net.sourceforge.myvd.server.Server.main(Server.java:534) [myvd-server-1.0.1.jar:?]". 

Again, I've tried several different configurations, but nothing seems to address this message as long as VirtualMemberOf is included in the chain.

Some additional details about the environment:
My Directory's baseDN = dc=home,dc=domain,dc=net
The directory is running on the localhost, the firewall is off

Here is my full myvd.conf
________________________________________________________________________

#################################################

#Configure global chains
#server.globalChain=accessLog,LogAllTransactions
server.globalChain=accessLog,cache,LogAllTransactions

#Configure Logging insert
server.globalChain.accessLog.className=net.sourceforge.myvd.inserts.AccessLog
server.globalChain.accessLog.config.fileName=%MYVD_HOME%/logs/myvd.access.log

#Configure Caching insert
server.globalChain.cache.className=net.sourceforge.myvd.inserts.cache.simple.SimpleCache

#Configure LogAllTransactions insert
server.globalChain.LogAllTransactions.className=net.sourceforge.myvd.inserts.DumpTransaction
server.globalChain.LogAllTransactions.config.logLevel=debug
server.globalChain.LogAllTransactions.config.label=ATW

#################################################

#Establish local namespaces
server.nameSpaces=Root,FreeIPA

#Root Namespace
server.Root.chain=RootDSE
server.Root.nameSpace=
server.Root.weight=100

#RootDSE insert
server.Root.RootDSE.className=net.sourceforge.myvd.inserts.RootDSE
server.Root.RootDSE.config.namingContexts=cn=changelog|dc=home,dc=domain,dc=net|o=ipaca

#FreeIPA LDAP Namespace
server.FreeIPA.chain=LDAPBaseServer
#server.FreeIPA.chain=VirtualMemberOf,LDAPBaseServer
server.FreeIPA.nameSpace=dc=home,dc=domain,dc=net
server.FreeIPA.weight=0

#VirtualMemberOf insert
server.FreeIPA.VirtualMemberOf.className=net.sourceforge.myvd.inserts.mapping.VirtualMemberOf
server.FreeIPA.VirtualMemberOf.searchBase=cn=groups,cn=accounts,dc=home,dc=domain,dc=net
server.FreeIPA.VirtualMemberOf.applyToObjectClass=inetOrgPerson
server.FreeIPA.VirtualMemberOf.attributeName=vmemberof
server.FreeIPA.VirtualMemberOf.searchObjectClass=groupOfNames
server.FreeIPA.VirtualMemberOf.searchAttribute=member
server.FreeIPA.VirtualMemberOf.replace=true

#LDAPBaseServer insert
server.FreeIPA.LDAPBaseServer.className=net.sourceforge.myvd.inserts.ldap.LDAPInterceptor
server.FreeIPA.LDAPBaseServer.config.host=localhost
server.FreeIPA.LDAPBaseServer.config.port=389
server.FreeIPA.LDAPBaseServer.config.remoteBase=dc=home,dc=domain,dc=net

_________________________________________________________

I greatly appreciate any direction or feedback you can offer!

Best regards,
Drew

**Comment by mlbiam** [comment]
@drew504 

> I have an application that will only read attributes from the inetOrgPerson objectclass. I would like to leverage the memberOf attribute within my application, but in my directory (389/FreeIPA) memberOf lives in the inetUser objectclass. As a result, memberOf is invisible to my application.

I'm not sure I understand the issue you are having.  You can't read attributes that aren't part of the inetOrgPerson objectClass?  There's nothing in LDAP that would enforce that.  Can you share the application you are using?  Is the issue that it only knows to look at certain attributes and you just need to make sure that your memberof is named correctly?  I see what you're saying about attribute being in inetuser, but honestly I can't see a way to enforce that you can't read an attribute other then a access control rule?

Also, is your only directory this FreeIPA?  If so then the first question I think we should solve is why can't you see the memberOf attribute because it could be as simple as a configuration change on your application and you don't need MyVD or a simpler approach can be used of just renaming the attribute in the proxy instead of recreating memberof in MyVD.

> "Missing schema location in RootDSE, using default schema." - First thing I did was configure a simple proxy following the directions. It works, but when I connect with Apache Directory Studio, I get the above error. I'm not sure if this is benign or indicative of a larger issue. - I tried different configurations with the RootDSE and RootObject inserts. Neither seemed to have any impact.

This is because MyVD doesn't maintain any internal schema.  Its usually not needed.  Apache Directory Studio and other browsers like it to make it easier to interact with a directory.  The workaround is to proxy the remote directory (in this case 389) schema.  I can give you a config that will do that assuming we still need MyVD.

> When I include the VirtualMemberOf insert in my namespace chain, the application will not start and prints the following in the logs:
"[2017-10-09 16:25:37,656][main] ERROR Server - Error starting server : java.lang.IllegalArgumentException: Attribute value cannot be null
java.lang.IllegalArgumentException: Attribute value cannot be null
at com.novell.ldap.LDAPAttribute.addValue(LDAPAttribute.java:227) ~[jldap-1.0.0.jar:?]
at net.sourceforge.myvd.inserts.mapping.VirtualMemberOf.configure(VirtualMemberOf.java:90) ~[myvd-server-1.0.1.jar:?]
at net.sourceforge.myvd.core.InsertChain.configureChain(InsertChain.java:192) ~[myvd-server-1.0.1.jar:?]
at net.sourceforge.myvd.server.ServerCore.configureChain(ServerCore.java:80) ~[myvd-server-1.0.1.jar:?]
at net.sourceforge.myvd.server.ServerCore.buildNamespaces(ServerCore.java:150) ~[myvd-server-1.0.1.jar:?]
at net.sourceforge.myvd.server.ServerCore.startService(ServerCore.java:169) ~[myvd-server-1.0.1.jar:?]
at net.sourceforge.myvd.server.Server.startServer(Server.java:259) ~[myvd-server-1.0.1.jar:?]
at net.sourceforge.myvd.server.Server.main(Server.java:534) [myvd-server-1.0.1.jar:?]".

You need to add `.config.` to your virtualmeberof config for everything except `className`:

```
#VirtualMemberOf insert
server.FreeIPA.VirtualMemberOf.className=net.sourceforge.myvd.inserts.mapping.VirtualMemberOf
server.FreeIPA.VirtualMemberOf.config.searchBase=cn=groups,cn=accounts,dc=home,dc=domain,dc=net
server.FreeIPA.VirtualMemberOf.config.applyToObjectClass=inetOrgPerson
server.FreeIPA.VirtualMemberOf.config.attributeName=vmemberof
server.FreeIPA.VirtualMemberOf.config.searchObjectClass=groupOfNames
server.FreeIPA.VirtualMemberOf.config.searchAttribute=member
server.FreeIPA.VirtualMemberOf.config.replace=true
```

**Comment by drew504** [comment]
@mlbiam 

> I'm not sure I understand the issue you are having. You can't read attributes that aren't part of the inetOrgPerson objectClass? There's nothing in LDAP that would enforce that. Can you share the application you are using? Is the issue that it only knows to look at certain attributes and you just need to make sure that your memberof is named correctly? I see what you're saying about attribute being in inetuser, but honestly I can't see a way to enforce that you can't read an attribute other then a access control rule?

I'm working with an internally developed application. You pass it an objectClass (one and only one objectClass), it reads the schema to see what attributes the objectClass provides, then the UI lets you select attributes from that list to be returned in all future searches. inetOrgPerson provides all the attributes I'm interested in, with the exception of memberOf. This is why I'm hoping to map memberOf (or a variant like vmemberOf) into inetOrgPerson.

If the above makes sense and is doable, then I would like to please request the config to proxy the schema for 389.

I'm feeling pretty silly about the .config. miss. Sorry for wasting your time on that one.

Best regards,
Drew

**Comment by mlbiam** [comment]
> I'm working with an internally developed application. You pass it an objectClass (one and only one objectClass), it reads the schema to see what attributes the objectClass provides, then the UI lets you select attributes from that list to be returned in all future searches. inetOrgPerson provides all the attributes I'm interested in, with the exception of memberOf. This is why I'm hoping to map memberOf (or a variant like vmemberOf) into inetOrgPerson.

Ahh, I see.  So the issue isn't that you need to generate memberof, you need your app to think its part of inetOrgPerson due to the schema.  hmm,  i think we can do that with some cleverness.  I've got FreeIPA running in my lab so let me see what we can do....

**Comment by mlbiam** [comment]
@drew504 here's a config to try.  It maps the schema entry for inetOrgPerson to include the memberOf attribute.  You'll need to update it for your environment:

```
server.listener.port=10983
#Configure global chains
server.globalChain=


server.nameSpaces=Root,schema,freeipa

#Define RootDSE
server.Root.chain=RootDSE
server.Root.nameSpace=
server.Root.weight=0
server.Root.RootDSE.className=net.sourceforge.myvd.inserts.RootDSE
server.Root.RootDSE.config.namingContexts=dc=rhelent,dc=lan
server.Root.RootDSE.config.supportedControls=2.16.840.1.113730.3.4.18,2.16.840.1.113730.3.4.2,1.3.6.1.4.1.4203.1.10.1,1.2.840.113556.1.4.319,1.2.826.0.1.334810.2.3,1.2.826.0.1.3344810.2.3,1.3.6.1.1.13.2,1.3.6.1.1.13.1,1.3.6.1.1.12
server.Root.RootDSE.config.supportedSaslMechanisms=NONE
server.Root.RootDSE.config.subSchemaSubEntry=cn=schema

server.schema.chain=addmemberof,ldap
server.schema.nameSpace=cn=schema
server.schema.weight=0
server.schema.addmemberof.className=net.sourceforge.myvd.inserts.mapping.AttributeValueMapper
server.schema.addmemberof.config.mapping=objectClasses.( 2.16.840.1.113730.3.2.2 NAME 'inetOrgPerson' SUP organizationalPerson STRUCTURAL MAY ( memberOf $ audio $ businessCategory $ carLicense $ departmentNumber $ displayName $ employeeNumber $ employeeType $ givenName $ homePhone $ homePostalAddress $ initials $ jpegPhoto $ labeledURI $ mail $ manager $ mobile $ o $ pager $ photo $ roomNumber $ secretary $ uid $ userCertificate $ x500UniqueIdentifier $ preferredLanguage $ userSMIMECertificate $ userPKCS12 ) X-ORIGIN 'RFC 2798' )=( 2.16.840.1.113730.3.2.2 NAME 'inetOrgPerson' SUP organizationalPerson STRUCTURAL MAY ( audio $ businessCategory $ carLicense $ departmentNumber $ displayName $ employeeNumber $ employeeType $ givenName $ homePhone $ homePostalAddress $ initials $ jpegPhoto $ labeledURI $ mail $ manager $ mobile $ o $ pager $ photo $ roomNumber $ secretary $ uid $ userCertificate $ x500UniqueIdentifier $ preferredLanguage $ userSMIMECertificate $ userPKCS12 ) X-ORIGIN 'RFC 2798' )
server.schema.ldap.className=net.sourceforge.myvd.inserts.ldap.LDAPInterceptor
server.schema.ldap.config.host=freeipa.rhelent.lan
server.schema.ldap.config.port=389
server.schema.ldap.config.remoteBase=cn=schema
server.schema.ldap.config.minimumConnections=10
server.schema.ldap.config.maximumConnections=10
server.schema.ldap.config.type=ldap
server.schema.ldap.config.ignoreRefs=true

server.freeipa.chain=ldap
server.freeipa.nameSpace=dc=rhelent,dc=lan
server.freeipa.wieght=0
server.freeipa.ldap.className=net.sourceforge.myvd.inserts.ldap.LDAPInterceptor
server.freeipa.ldap.config.host=freeipa.rhelent.lan
server.freeipa.ldap.config.port=389
server.freeipa.ldap.config.remoteBase=dc=rhelent,dc=lan
server.freeipa.ldap.config.minimumConnections=10
server.freeipa.ldap.config.maximumConnections=10
server.freeipa.ldap.config.type=ldap
server.freeipa.ldap.config.ignoreRefs=true
server.freeipa.ldap.config.passBindOnly=false
```
NOTE: if the schema changes for inetOrgPeson (which it shouldn't) this would stop working

OH, and don't worry about the `.config.` thing, I wrote this system and I do it all the time!

**Comment by drew504** [comment]
Yes, You've got it! It is not about generating memberOf, it is about making it a part of the inetOrgPerson schema. 

I just tried the config above. It resolved my RootDSE error, I am now able to connect using ApacheDS without any errors. Unfortunately, it doesn't look like it is overriding the inetOrgPerson schema.

My application connects OK and returns the inetOrgPerson schema attributes, but without memberOf (also tested with ApacheDS, same result). 

Possibly unrelated, but I also noticed that when I connect with my application I get the following error in the myvd log "ERROR BindRequestHandler - ERR_160 Bind error : DIGEST-MD5 mechanism not supported. Please check the server.xml configuration file (supportedMechanisms field)". I don't see this or any log entries when I bind using ApacheDS.

Unfortunately I'll be on the road all day today. I will try to dig in deeper, double-check my config, and provide more detail this evening or tomorrow morning.

Thank you and Best regards,
Drew

**Comment by mlbiam** [comment]
> My application connects OK and returns the inetOrgPerson schema attributes, but without memberOf (also tested with ApacheDS, same result).

See if ApacheDS is rechecking for your schema.  You can also try an ldapsearch:

```
$ ldapsearch -x -h localhost -p 10983 -b 'cn=schema' -s sub -o ldif-wrap=no '(objectClass=*)' '+' | grep inetOrgPerson
```
it should look like:
```
objectClasses: ( 2.16.840.1.113730.3.2.2 NAME 'inetOrgPerson' SUP organizationalPerson STRUCTURAL MAY ( memberOf $ audio $ businessCategory $ carLicense $ departmentNumber $ displayName $ employeeNumber $ employeeType $ givenName $ homePhone $ homePostalAddress $ initials $ jpegPhoto $ labeledURI $ mail $ manager $ mobile $ o $ pager $ photo $ roomNumber $ secretary $ uid $ userCertificate $ x500UniqueIdentifier $ preferredLanguage $ userSMIMECertificate $ userPKCS12 ) X-ORIGIN 'RFC 2798' )
```
if, it doesn't, please add the output to the ticket.

The second issue is a harder one.  Its trying to use MD5 digest instead of generic LDAP Bind operations.  The idea is that you don't have to send the actual password, just the hash.  Of course MD5 is outdated and no one uses it, or digest authentication anymore so MyVD doesn't support it.  Can you tell it to use a generic LDAP Bind instead?

**Comment by drew504** [comment]
OK, I think I've got this working properly now.

It didn't look like the AttributeValueMapper insert was getting triggered. So, I took the ldapsearch command you provided above and ran it directly against my directory. My output for the inetOrgPerson schema was slightly different than yours so I don't think it was triggering the insert. Here is the output from my directory.
`objectClasses: ( 2.16.840.1.113730.3.2.2 NAME 'inetOrgPerson' SUP organizationalPerson STRUCTURAL MAY ( audio $ businessCategory $ carLicense $ departmentNumber $ displayName $ employeeNumber $ employeeType $ givenName $ homePhone $ homePostalAddress $ initials $ jpegPhoto $ labeledURI $ mail $ manager $ mobile $ o $ pager $ photo $ roomNumber $ secretary $ uid $ userCertificate $ x500UniqueIdentifier $ preferredLanguage $ userSMIMECertificate $ userPKCS12 ) X-ORIGIN ( 'RFC 2798' 'user defined' ) )`

I took that ouput and updated the AttributeValueMapper line to this:

`server.schema.addmemberof.config.mapping=objectClasses.( 2.16.840.1.113730.3.2.2 NAME 'inetOrgPerson' SUP organizationalPerson STRUCTURAL MAY ( audio $ businessCategory $ carLicense $ departmentNumber $ displayName $ employeeNumber $ employeeType $ givenName $ homePhone $ homePostalAddress $ initials $ jpegPhoto $ labeledURI $ mail $ manager $ memberOf $ mobile $ o $ pager $ photo $ roomNumber $ secretary $ uid $ userCertificate $ x500UniqueIdentifier $ preferredLanguage $ userSMIMECertificate $ userPKCS12 ) X-ORIGIN ( 'RFC 2798' 'user defined' ) )=( 2.16.840.1.113730.3.2.2 NAME 'inetOrgPerson' SUP organizationalPerson STRUCTURAL MAY ( audio $ businessCategory $ carLicense $ departmentNumber $ displayName $ employeeNumber $ employeeType $ givenName $ homePhone $ homePostalAddress $ initials $ jpegPhoto $ labeledURI $ mail $ manager $ mobile $ o $ pager $ photo $ roomNumber $ secretary $ uid $ userCertificate $ x500UniqueIdentifier $ preferredLanguage $ userSMIMECertificate $ userPKCS12 ) X-ORIGIN ( 'RFC 2798' 'user defined' ) )`

Now when I run the ldapsearch it returns the schema with memberOf included!

At this point I started testing with my application. Initially the memberOf values were returned and everything worked great. Then some strange behavior started. Subsequent searches were not returning memberOf until I restarted my app. As near as I can figure out, MyVD didn't like something about the way my application is managing session or sending requests. Maybe this is somehow related to the MD5 error message. After my application performs the initial search I see the following in the myvd.log.
`LdapSession - AbandonableRequest with messageId 40 not found in outstandingRequests.` It seems like all the searches after this message are acting as an anonymous bind and thus cannot access the memberOf attribute.

At this point I configured a proxy username and password, and set passBindOnly=true.

Now the behavior is consistent and my application is working reliably!

Next I will be looking to setup a key store and transition to a secure connection. But that is a challenge for another day.

Thank you for the fantastic support!

Best regards,
Drew
